game/kernel/asm_funcs_arm64.s

```diff
@@ -0,0 +1,262 @@
+  ;; sub sp, 8
+  ldr q0, [sp, #+16]
+  ;; grab the stack offset
+  ldr x0, [sp, #+8]
+
+  ;; first, save quadword registers
+  stp q15, q14, [sp, #-32]!
+  stp q13, q12, [sp, #-32]!
+  stp q11, q10, [sp, #-32]!
+  stp q9, q8, [sp, #-32]!
+
+  ; NOTE - in x86 the 2 special registers are saved (R10 and R11)
+  ; we don't need to do that in ARM64, there are plenty of registers to work with
+
+  ;; oof
+  sub sp, sp, 1280
+  str x0, [sp, #+64] ; arg 0 (RDI in x86) and 
+  str x1, [sp, #+80] ; arg 1 (RSI in x86)
+  str x2, [sp, #+96] ; arg 2 (RDX in x86) and arg 3 (RCX in x86)
+  str x3, [sp, #+112] ; arg 2 (RDX in x86) and arg 3 (RCX in x86)
+  str x4, [sp, #+128] ; arg 4 (R8 in x86) and arg 5 (R8 in x86)
+  str x5, [sp, #+144] ; arg 4 (R8 in x86) and arg 5 (R8 in x86)
+  str x6, [sp, #+160] ; arg 6 (R10 in x86) and arg 7 (R11 in x86)
+  str x7, [sp, #+176] ; arg 6 (R10 in x86) and arg 7 (R11 in x86)
+  str x20, [sp, #+352] ;; s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
+  str x21, [sp, #+368] ;; s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
+
+  mov x0, sp ; move the stack pointer to arg 0
+  sub x0, x0, x22 ; R15 is a "special" offset TODO - whats special about it?
+  str x0, [sp, #+464] ;; mip2c code's MIPS stack
+
+  mov x0, sp ;; move the stack pointer to the new position
+
+  sub sp, sp, x8 ;; allocate space on the stack for GOAL fake stack
+  stp x8, x8, [sp, #-16]! ;; and remember this so we can find our way back
+
+  ;; TODO - this used to be a movq rax, xmm0
+  ;; TODO - not sure why an `xmm` was used because that movq only uses the lower 64bits anyway
+  mov x0, v0.d[0] ; represents the lower 64 bits of q0
+  blr x8 ;; call!
+
+  ;; unallocate
+  ldp x8, x8, [sp], #16
+  add sp, sp, x8
+
+  ldr x8, [sp, #+32]
+
+  add sp, sp, 1280 ; reset the stackpointer back
+
+  ldp x8, x9, [sp], #32
+  ldp x10, x11, [sp], #32
+  ldp x12, x13, [sp], #32
+  ldp x14, x15, [sp], #32
+
+  add sp, sp, 24 ;; 16 for the stuff pushed by trampoline
+  ret
+
+;; The _call_goal_asm function is used to call a GOAL function from C.
+;; It calls on the parent stack, which is a bad idea if your stack is not already a GOAL stack.
+;; It supports up to 3 arguments and a return value.
+;; This should be called with the arguments:
+;; - first goal arg
+;; - second goal arg
+;; - third goal arg
+;; - address of function to call
+;; - address of the symbol table
+;; - GOAL memory space offset
+.global _call_goal_asm_arm64
+.align 4
+_call_goal_asm_arm64:
+  ;; saved registers we need to modify for GOAL should be preserved
+  ; ARM64 requires 16-byte stack pointer alignment
+  stp x20, x21, [sp, #-16]!
+  str x22, [sp, #-16]!
+
+  ;; x0 - first arg
+  ;; x1 - second arg
+  ;; x2 - third arg
+  ;; x3 - function pointer
+  ;; x4 - st (goes in x20 and x21)
+  ;; x5 - off (goes in x22)
+
+  ;; set GOAL process
+  mov x20, x4
+  ;; symbol table
+  mov x21, x4
+  ;; offset
+  mov x22, x5
+  ;; call GOAL by function pointer
+  blr x3
```

Comment:
this ends up writing the `x30` (link register) with the address of the next instruction so the function being called can return to the right place.

Reply:
You'll need to back up this register and restore it before `ret` in this function, otherwise the `ret` and the end of this function will branch back to right after this line. (equivalent to how `ra` works in mips)